compiler/src/iree/compiler/Codegen/Common/test/canonicalize_early_bufferization_ops.mlir

```diff
@@ -0,0 +1,178 @@
+  iree_codegen.store_to_memref %expanded, %subview : tensor<3x3x96xf32> into memref<3x3x96xf32, strided<[288, 96, 1], offset: 864>, #hal.descriptor_type<storage_buffer>>
+  return
+}
+// CHECK-LABEL: @fold_reshape_with_slice_store
+//   CHECK-DAG:   %[[DEST_SUBSPAN:.+]] = hal.interface.binding.subspan{{.*}} binding(1)
+//       CHECK:   %[[SUBVIEW:.+]] = memref.subview %[[DEST_SUBSPAN]]
+//  CHECK-SAME:     memref<3x3x96xf32, strided<[288, 96, 1], offset: 864>
+//       CHECK:   %[[COLLAPSE:.+]] = memref.collapse_shape %[[SUBVIEW]]
+//  CHECK-SAME:     into memref<864xf32, strided<[1], offset: 864>
+//       CHECK:   %[[EXPAND:.+]] = memref.expand_shape %[[COLLAPSE]]
+//  CHECK-SAME:     into memref<3x3x1x96xf32, strided<[288, 96, 96, 1], offset: 864>
+//       CHECK:   iree_codegen.store_to_memref {{.*}}, %[[EXPAND]]
+
+// -----
+
+#pipeline_layout = #hal.pipeline.layout<constants = 4, bindings = [
+  #hal.pipeline.binding<storage_buffer>,
+  #hal.pipeline.binding<storage_buffer>
+]>
+func.func @fold_dynamic_reshape_load() {
+  %c0 = arith.constant 0 : index
+  %0 = hal.interface.constant.load layout(#pipeline_layout) ordinal(0) : index
+  %1 = hal.interface.constant.load layout(#pipeline_layout) ordinal(1) : index
+  %2 = hal.interface.constant.load layout(#pipeline_layout) ordinal(2) : index
+  %3 = hal.interface.constant.load layout(#pipeline_layout) ordinal(3) : index
+  %4 = hal.interface.binding.subspan layout(#pipeline_layout) binding(0) : memref<?x?xf32, #hal.descriptor_type<storage_buffer>>{%0, %1}
+  memref.assume_alignment %4, 1 : memref<?x?xf32, #hal.descriptor_type<storage_buffer>>
+  %5 = hal.interface.binding.subspan layout(#pipeline_layout) binding(1) : memref<?x?xf32, #hal.descriptor_type<storage_buffer>>{%2, %3}
+  memref.assume_alignment %5, 1 : memref<?x?xf32, #hal.descriptor_type<storage_buffer>>
+  %6 = iree_codegen.load_from_memref %4 : memref<?x?xf32, #hal.descriptor_type<storage_buffer>> -> tensor<?x?xf32>
+  %collapsed = tensor.collapse_shape %6 [[0, 1]] : tensor<?x?xf32> into tensor<?xf32>
+  %expanded = tensor.expand_shape %collapsed [[0, 1]] output_shape [%2, %3] : tensor<?xf32> into tensor<?x?xf32>
+  %barrier = util.optimization_barrier %expanded : tensor<?x?xf32>
+  iree_codegen.store_to_memref %barrier, %5 : tensor<?x?xf32> into memref<?x?xf32, #hal.descriptor_type<storage_buffer>>
+  return
+}
+// CHECK-LABEL: @fold_dynamic_reshape_load
+//   CHECK-DAG:   %[[D0:.+]] = hal.interface.constant.load{{.*}} ordinal(2) : index
+//   CHECK-DAG:   %[[D1:.+]] = hal.interface.constant.load{{.*}} ordinal(3) : index
+//   CHECK-DAG:   %[[SRC_SUBSPAN:.+]] = hal.interface.binding.subspan{{.*}} binding(0)
+//   CHECK-DAG:   %[[COLLAPSE:.+]] = memref.collapse_shape %[[SRC_SUBSPAN]]
+//   CHECK-DAG:   %[[EXPAND:.+]] = memref.expand_shape %[[COLLAPSE]]
+//  CHECK-SAME:     output_shape [%[[D0]], %[[D1]]]
+//   CHECK-DAG:   %[[LOAD:.+]] = iree_codegen.load_from_memref %[[EXPAND]]
```

Comment:
maybe add the types here as well so its easier for the reader to see the resulting IR?